From ff3f81b1bbbcd77b3125a81f91906c9d0802f9a9 Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Mon, 18 May 2026 07:46:09 +0000
Subject: [PATCH] pstore/ram: Fix crash when setting number of cpus to an odd
 number

jira VULN-4502
cve CVE-2023-52619
commit-author Weichen Chen <weichen.chen@mediatek.com>
commit d49270a04623ce3c0afddbf3e984cb245aa48e9c

When the number of cpu cores is adjusted to 7 or other odd numbers,
the zone size will become an odd number.
The address of the zone will become:
    addr of zone0 = BASE
    addr of zone1 = BASE + zone_size
    addr of zone2 = BASE + zone_size*2
    ...
The address of zone1/3/5/7 will be mapped to non-alignment va.
Eventually crashes will occur when accessing these va.

So, use ALIGN_DOWN() to make sure the zone size is even
to avoid this bug.

	Signed-off-by: Weichen Chen <weichen.chen@mediatek.com>
	Reviewed-by: Matthias Brugger <matthias.bgg@gmail.com>
	Tested-by: "Guilherme G. Piccoli" <gpiccoli@igalia.com>
Link: https://lore.kernel.org/r/20230224023632.6840-1-weichen.chen@mediatek.com
	Signed-off-by: Kees Cook <keescook@chromium.org>
(cherry picked from commit d49270a04623ce3c0afddbf3e984cb245aa48e9c)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 fs/pstore/ram.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/fs/pstore/ram.c b/fs/pstore/ram.c
index fefe3d391d3a..56a04fef5190 100644
--- a/fs/pstore/ram.c
+++ b/fs/pstore/ram.c
@@ -519,6 +519,7 @@ static int ramoops_init_przs(const char *name,
 	}
 
 	zone_sz = mem_sz / *cnt;
+	zone_sz = ALIGN_DOWN(zone_sz, 2);
 	if (!zone_sz) {
 		dev_err(dev, "%s zone size == 0\n", name);
 		goto fail;
-- 
2.47.3

